### PR TITLE
replace filter method from fuzzaldrin with native filter method of javascript to fix country select issue #2751

### PR DESCRIPTION
**Pull request by endo-aki22**
I want to merge this change because filter method of fuzzaldrin doesn't work properly in `SingleAutocompleteSelectField` component.
There is already an issue(#2751) related to this one.

### Screenshots
after changes..
Screenshot from 2022-12-02 00-12-02
Screenshot from 2022-12-02 00-12-15

### Pull Request Checklist
This code doesn't contain any UI changes.
And this code meets all requirements above.

1. [ ] This code contains UI changes
2. [ ] All visible strings are translated with proper context including data-formatting
3. [ ] Attributes `[data-test-id]` are added for new elements
4. [ ] Changes are mentioned in the changelog
5. [ ] The changes are tested in different browsers and in light/dark mode

### Test environment config

API_URI=https://automation-dashboard.staging.saleor.cloud/graphql/
MARKETPLACE_URL=https://apps.saleor.io
SALEOR_APPS_ENDPOINT=https://apps.saleor.io/api/saleor-apps

### Do you want to run more stable tests?
To run all tests, just select the stable checkbox. To speed up tests, increase the number of containers. Tests will be re-run only when the "run e2e" label is added.

1. [ ] stable
2. [ ] giftCard
3. [ ] category
4. [ ] collection
5. [ ] attribute
6. [ ] productType
7. [ ] shipping
8. [ ] customer
9. [ ] permissions
10. [ ] menuNavigation
11. [ ] pages
12. [ ] sales
13. [ ] vouchers
14. [ ] homePage
15. [ ] login
16. [ ] orders
17. [ ] products
18. [ ] app
19. [ ] plugins
20. [ ] translations
21. [ ] navigation
22. [ ] variants
23. [ ] payments

CONTAINERS=1

### CODE REVIEW

**Review by Droniu — APPROVED** [review]
Thanks for the contribution! :)

**Comment by andrzejewsky** [comment]
@endo-aki22 Hey, could you use https://fusejs.io/ ? We would love to migrate from .`fuzzaldrin` - Having simple js filtering (as it is right now) removes fuzzy search and think this file can be the first place that will adopt fusejs.

**Comment by endo-aki22** [comment]
understood. I will try it.

**Comment by endo-aki22** [comment]
I added fuse.js, and replaced filter of `fuzzaldrin` with search `fuse.js`.
It works well as we expected.
Please check it.

**Comment by krzysztofzuraw** [comment]
@endo-aki22 can you please rebase with main branch

**Comment by endo-aki22** [comment]
> @endo-aki22 can you please rebase with main branch

sure

**Comment by endo-aki22** [comment]
> @endo-aki22 can you please rebase with main branch

done

**Review by szczecha — DISMISSED** [review]

File: src/components/SingleAutocompleteSelectField/SingleAutocompleteSelectField.tsx [hunk 1]
```diff
@@ -298,12 +298,12 @@ const SingleAutocompleteSelectField: React.FC<SingleAutocompleteSelectFieldProps
...
 
+  const fuse = new Fuse(choices, { useExtendedSearch: true, keys: ["label"] });
+
   return (
     <SingleAutocompleteSelectFieldComponent
       fetchChoices={q => setQuery(q || "")}
-      choices={filter(choices, query, {
-        key: "label",
-      })}
+      choices={fuse.search(`'${query}`).map(v => v.item)}
```
[the remark this answers is not in the document]
Comment (endo-aki22) [reply src/components/SingleAutocompleteSelectField/SingleAutocompleteSelectField.tsx:306]: > isn't it the same? Why do you actually add additional `'` ?

it's different, and different result.
using `'query`, the result include the `query`.
https://fusejs.io/examples.html#extended-search

[the remark this answers is not in the document]
Comment (endo-aki22) [reply src/components/SingleAutocompleteSelectField/SingleAutocompleteSelectField.tsx]: > yes. removed `useExtendedSearch` and include-match `'`. please check it now.

**Comment by szczecha** [comment]
@saleor/dashboard @endo-aki22  I have run the e2e test and did some testing around `selects` in the whole app, and I'm afraid that we losing some functionality here.

Before this change when I click on select I had a list of available options e.g tax classes, currencies, countries, warehouses in cancel fulfillment dialog, also I was able to use search to find and choose my option. 

https://user-images.githubusercontent.com/30683248/207388309-38b5d608-d1f2-4770-a45f-c8dab283376e.mov
On left our Demo on right localhost build from this branch

**Comment by endo-aki22** [comment]
> @saleor/dashboard @endo-aki22 I have run the e2e test and did some testing around `selects` in the whole app, and I'm afraid that we losing some functionality here.
> 
> Before this change when I click on select I had a list of available options e.g tax classes, currencies, countries, warehouses in cancel fulfillment dialog, also I was able to use search to find and choose my option.
> 
> https://user-images.githubusercontent.com/30683248/207388309-38b5d608-d1f2-4770-a45f-c8dab283376e.mov On left our Demo on right localhost build from this branch

fusejs results an empty array if search query is an empty string. so I fixed it to use original array if query is empty. please check it.

**Comment by szczecha** [comment]
It's better now, all e2e fails related to invisible options for select now pass.

File: src/components/SingleAutocompleteSelectField/SingleAutocompleteSelectField.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (endo-aki22) [reply src/components/SingleAutocompleteSelectField/SingleAutocompleteSelectField.tsx]: Hi, @andrzejewsky. please check.

**Comment by endo-aki22** [comment]
> It's better now, all e2e fails related to invisible options for select now pass.

**Comment by endo-aki22** [comment]
> It's better now, all e2e fails related to invisible options for select now pass.

I accidentally clicked the close button. how can we revert it?

File: src/components/SingleAutocompleteSelectField/SingleAutocompleteSelectField.tsx [hunk 2]
```diff
@@ -298,12 +298,12 @@ const SingleAutocompleteSelectField: React.FC<SingleAutocompleteSelectFieldProps
...
 
+  const fuse = new Fuse(choices, { keys: ["label"] });
+
   return (
     <SingleAutocompleteSelectFieldComponent
       fetchChoices={q => setQuery(q || "")}
-      choices={filter(choices, query, {
-        key: "label",
-      })}
+      choices={query !== "" ? fuse.search(query).map(v => v.item) : choices}
```
Comment (andrzejewsky) [inline src/components/SingleAutocompleteSelectField/SingleAutocompleteSelectField.tsx:306]: I will give you approval anyway, but if you have time, please consider now a more nice pattern - How about wrapping this into react hook, so we can re-use it across the other views. You can call the hook eg. `useFuzzySearch` that will create an instance of fusejs and share proper functions and results, encapsulating also the details like this map here.


```ts
const useFuzzySearch = <T>(items: T, keys: string[]) => {
  const fuse = new Fuse(items, { keys });
  
  const search = (query: string) => {
    if (!query) return items;
  
    return fuse.search(query)
  }
 
  return { search }
 }

```

Comment (andrzejewsky) [reply src/components/SingleAutocompleteSelectField/SingleAutocompleteSelectField.tsx:306]: now we have an abstraction to it, and we can even replace fuzzy search library in one place

**Review by andrzejewsky — DISMISSED** [review]

**Review by szczecha — APPROVED** [review]

File: src/components/SingleAutocompleteSelectField/SingleAutocompleteSelectField.tsx (hunk 2 above)
Comment (endo-aki22) [reply src/components/SingleAutocompleteSelectField/SingleAutocompleteSelectField.tsx:306]: > now we have an abstraction to it, and we can even replace fuzzy search library in one place

you are right. if needed, and used it other components, I will do it later.

**Review by Droniu — DISMISSED** [review]

File: src/components/SingleAutocompleteSelectField/SingleAutocompleteSelectField.tsx (hunk 2 above)
Comment (andrzejewsky) [reply src/components/SingleAutocompleteSelectField/SingleAutocompleteSelectField.tsx:306]: OK, do you want to change this or anything else now or we can merge it? (We'll do this abstraction anyway once we start migrating to fuse in the other places)

Comment (endo-aki22) [reply src/components/SingleAutocompleteSelectField/SingleAutocompleteSelectField.tsx:306]: what do you mean?

Comment (endo-aki22) [reply src/components/SingleAutocompleteSelectField/SingleAutocompleteSelectField.tsx:306]: I just want you to merge it.

Comment (endo-aki22) [reply src/components/SingleAutocompleteSelectField/SingleAutocompleteSelectField.tsx:306]: Shall I rebase with `main`?

**Review by Droniu — APPROVED** [review]

**Review by szczecha — APPROVED** [review]

**Review by andrzejewsky — APPROVED** [review]

**Pull request merged**